File: unit-spa/backend/src/main/java/otus/amogilevskiy/spring/service/genre/GenreService.java
package otus.amogilevskiy.spring.service.genre;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import otus.amogilevskiy.spring.domain.Genre;
import otus.amogilevskiy.spring.dto.genre.GenreDto;

public interface GenreService {

    Genre create(GenreDto dto);

    Genre updateById(long id, GenreDto dto);

    Page<Genre> findAll(Pageable pageable);

    Genre findById(long id);

}

File: unit-spa/frontend/src/services/GenreService.ts
import GenreDto from "../dtos/GenreDto";
import axios, {AxiosInstance} from "axios";
import Genre from "../domain/Genre";
import applyCaseMiddleware from "axios-case-converter";
import {PageRequest} from "./PageRequest";
import {PaginatedList} from "./PaginatedList";

class GenreService {

    private client: AxiosInstance;

    constructor(client: AxiosInstance) {
        this.client = client;
    }

    async save(dto: GenreDto): Promise<Genre> {
        let response;
        if (dto.id) {
            response = await this.client.patch(`/api/1.0/genres/${dto.id}`, dto);
        } else {
            response = await this.client.post('/api/1.0/genres', dto);
        }
        return response.data;
    }

    async getAll(request?: PageRequest): Promise<PaginatedList<Genre>> {
        const response = await this.client.get('/api/1.0/genres', {
            params: {
                ...request,
            },
        });
        return response.data;
    }

}

export default new GenreService(applyCaseMiddleware(axios.create()));
